src/poly/mod.rs

```diff
@@ -47,6 +47,60 @@ impl<F, V, M> Expr<F, V, M> {
             Expr::MI(_, _) => panic!("not implemented"),
         }
     }
+
+    pub fn meta(&self) -> &M {
+        match self {
+            Expr::Const(_, m) => m,
+            Expr::Sum(_, m) => m,
+            Expr::Mul(_, m) => m,
+            Expr::Neg(_, m) => m,
+            Expr::Pow(_, _, m) => m,
+            Expr::Query(_, m) => m,
+            Expr::Halo2Expr(_, m) => m,
+            Expr::MI(_, m) => m,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct HashResult {
+    pub hash: u64,
+    pub degree: usize,
+}
+
+impl<F: Field + Hash, V: Debug + Clone + Eq + Hash> Expr<F, V, ()> {
+    /// Uses Schwartz-Zippel Lemma to hash
+    pub fn hash(&self, assignments: &[VarAssignments<F, V>]) -> Expr<F, V, HashResult> {
```

Comment:
What do you think of passing a shared instance of hasher into all the expressions upon their creation? In that case we would be sure that all hashes are evaluated at the same point without the need to pass assignments as an argument.

Reply:
I think it makes sense, thanks for the observation 🙏🏼

Reply:
I'll close the discussion for now as we're not sure whether we would call the hash function in more than one place across the project. We can keep that in mind for the future.

Reply:
As discussed for now it will be implemented with the array of assignments, later on we can check if we need to use a var manager and pass the reference to the hasher instead